Project-Management/CoreTeam/workspace-notes.md

```diff
@@ -0,0 +1,43 @@
+# Workspace
+
+## Issues with the current workspace
+
+Focussing mainly on `MatrixWorkspace`.
+
+1. Constructed iteratively, i.e., workspaces start their life in an incomplete and dysfunctional state, vital information is added only later (size, instrument, spectrum to detector mapping).
+1. No encapsulation for data it contains, thus violating invariants:
+   - Access to X, Y, and E makes breaking or abusing internal data easy.
+   - Access to spectrum numbers on `ISpectrum` prevents validation (e.g., duplicates).
+   - Detector IDs can be set arbitrarily on `ISpectrum`, no check if this is in the instrument.
+1. `MatrixWorkspace` is meant to store histograms, but is used for many other purposes:
+   - Single values.
+   - Single data points in each histogram. Leads to performance problems due to massive overhead from using histograms for this purpose, e.g., at ILL.
+   - Masking.
```

Comment:
Defining a region-of-interest is another sore point in Mantid. Every algorithm has their own way of asking for and handling histogram and bin ranges, and you never know if your masking is respected or not. To be frank, the whole masking thing seems to be just an afterthought. Maybe something like a generalized 'selection' object which you could attach to a workspace to indicate which parts of the data, e.g. which histograms, which bins/X ranges should be touched by the algorithms would help here.

Reply:
"Every algorithm has their own way of asking for and handling histogram and bin ranges" -> see work by Lamar done for ESS: http://docs.mantidproject.org/nightly/concepts/IndexProperty.html. We now *do* have such a unified way, but it would need to be rolled out to algorithms.